File: lib/es6/gc.js


import * as Curry from "./curry.js";
import * as Printf from "./printf.js";
import * as Caml_gc from "./caml_gc.js";

var dummy_stat = {
  minor_words: 0,
  promoted_words: 0,
  major_words: 0,
  minor_collections: 0,
  major_collections: 0,
  heap_words: 0,
  heap_chunks: 0,
  live_words: 0,
  live_blocks: 0,
  free_words: 0,
  free_blocks: 0,
  largest_free: 0,
  fragments: 0,
  compactions: 0,
  top_heap_words: 0,
  stack_size: 0
};

function stat(param) {
  return dummy_stat;
}

function quick_stat(param) {
  return dummy_stat;
}

function get(param) {
  return {
          minor_heap_size: 0,
          major_heap_increment: 0,
          space_overhead: 0,
          verbose: 0,
          max_overhead: 0,
          stack_limit: 0,
          allocation_policy: 0,
          window_size: 0
        };
}

function print_stat(c) {
  var st = stat(undefined);
  Curry._1(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "minor_collections: ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: /* No_padding */0,
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "minor_collections: %d\n"
          }), st.minor_collections);
  Curry._1(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "major_collections: ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: /* No_padding */0,
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "major_collections: %d\n"
          }), st.major_collections);
  Curry._1(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "compactions:       ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: /* No_padding */0,
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "compactions:       %d\n"
          }), st.compactions);
  Printf.fprintf(c, /* Format */{
        _0: {
          TAG: /* Char_literal */12,
          _0: /* "\n" */10,
          _1: /* End_of_format */0
        },
        _1: "\n"
      });
  var l1 = Curry._1(Printf.sprintf(/* Format */{
            _0: {
              TAG: /* Float */8,
              _0: /* Float_f */0,
              _1: /* No_padding */0,
              _2: /* Lit_precision */{
                _0: 0
              },
              _3: /* End_of_format */0
            },
            _1: "%.0f"
          }), st.minor_words).length;
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "minor_words:    ",
              _1: {
                TAG: /* Float */8,
                _0: /* Float_f */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* Lit_precision */{
                  _0: 0
                },
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "minor_words:    %*.0f\n"
          }), l1, st.minor_words);
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "promoted_words: ",
              _1: {
                TAG: /* Float */8,
                _0: /* Float_f */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* Lit_precision */{
                  _0: 0
                },
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "promoted_words: %*.0f\n"
          }), l1, st.promoted_words);
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "major_words:    ",
              _1: {
                TAG: /* Float */8,
                _0: /* Float_f */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* Lit_precision */{
                  _0: 0
                },
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "major_words:    %*.0f\n"
          }), l1, st.major_words);
  Printf.fprintf(c, /* Format */{
        _0: {
          TAG: /* Char_literal */12,
          _0: /* "\n" */10,
          _1: /* End_of_format */0
        },
        _1: "\n"
      });
  var l2 = Curry._1(Printf.sprintf(/* Format */{
            _0: {
              TAG: /* Int */4,
              _0: /* Int_d */0,
              _1: /* No_padding */0,
              _2: /* No_precision */0,
              _3: /* End_of_format */0
            },
            _1: "%d"
          }), st.top_heap_words).length;
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "top_heap_words: ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "top_heap_words: %*d\n"
          }), l2, st.top_heap_words);
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "heap_words:     ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "heap_words:     %*d\n"
          }), l2, st.heap_words);
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "live_words:     ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "live_words:     %*d\n"
          }), l2, st.live_words);
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "free_words:     ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "free_words:     %*d\n"
          }), l2, st.free_words);
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "largest_free:   ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "largest_free:   %*d\n"
          }), l2, st.largest_free);
  Curry._2(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "fragments:      ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: {
                  TAG: /* Arg_padding */1,
                  _0: /* Right */1
                },
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "fragments:      %*d\n"
          }), l2, st.fragments);
  Printf.fprintf(c, /* Format */{
        _0: {
          TAG: /* Char_literal */12,
          _0: /* "\n" */10,
          _1: /* End_of_format */0
        },
        _1: "\n"
      });
  Curry._1(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "live_blocks: ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: /* No_padding */0,
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "live_blocks: %d\n"
          }), st.live_blocks);
  Curry._1(Printf.fprintf(c, /* Format */{
            _0: {
              TAG: /* String_literal */11,
              _0: "free_blocks: ",
              _1: {
                TAG: /* Int */4,
                _0: /* Int_d */0,
                _1: /* No_padding */0,
                _2: /* No_precision */0,
                _3: {
                  TAG: /* Char_literal */12,
                  _0: /* "\n" */10,
                  _1: /* End_of_format */0
                }
              }
            },
            _1: "free_blocks: %d\n"
          }), st.free_blocks);
  return Curry._1(Printf.fprintf(c, /* Format */{
                  _0: {
                    TAG: /* String_literal */11,
                    _0: "heap_chunks: ",
                    _1: {
                      TAG: /* Int */4,
                      _0: /* Int_d */0,
                      _1: /* No_padding */0,
                      _2: /* No_precision */0,
                      _3: {
                        TAG: /* Char_literal */12,
                        _0: /* "\n" */10,
                        _1: /* End_of_format */0
                      }
                    }
                  },
                  _1: "heap_chunks: %d\n"
                }), st.heap_chunks);
}

function allocated_bytes(param) {
  var match = Caml_gc.caml_gc_counters(undefined);
  return (match[0] + match[2] - match[1]) * 4;
}

function finalise_last(param, param$1) {
  
}

function call_alarm(arec) {
  if (arec.active.contents) {
    Caml_gc.caml_final_register(call_alarm, arec);
    return Curry._1(arec.f, undefined);
  }
  
}

function create_alarm(f) {
  var arec_active = {
    contents: true
  };
  var arec = {
    active: arec_active,
    f: f
  };
  Caml_gc.caml_final_register(call_alarm, arec);
  return arec_active;
}

function delete_alarm(a) {
  a.contents = false;
  
}

var finalise = Caml_gc.caml_final_register;

var finalise_release = Caml_gc.caml_final_release;

export {
  stat ,
  quick_stat ,
  get ,
  print_stat ,
  allocated_bytes ,
  finalise ,
  finalise_last ,
  finalise_release ,
  create_alarm ,
  delete_alarm ,
  
}
/* No side effect */

File: syntax/tests/idempotency/ocaml/stdlib/gc.ml
(**************************************************************************)
(*                                                                        *)
(*                                 OCaml                                  *)
(*                                                                        *)
(*             Damien Doligez, projet Para, INRIA Rocquencourt            *)
(*                                                                        *)
(*   Copyright 1996 Institut National de Recherche en Informatique et     *)
(*     en Automatique.                                                    *)
(*                                                                        *)
(*   All rights reserved.  This file is distributed under the terms of    *)
(*   the GNU Lesser General Public License version 2.1, with the          *)
(*   special exception on linking described in the file LICENSE.          *)
(*                                                                        *)
(**************************************************************************)

type stat = {
  minor_words : float;
  promoted_words : float;
  major_words : float;
  minor_collections : int;
  major_collections : int;
  heap_words : int;
  heap_chunks : int;
  live_words : int;
  live_blocks : int;
  free_words : int;
  free_blocks : int;
  largest_free : int;
  fragments : int;
  compactions : int;
  top_heap_words : int;
  stack_size : int;
}

type control = {
  mutable minor_heap_size : int;
  mutable major_heap_increment : int;
  mutable space_overhead : int;
  mutable verbose : int;
  mutable max_overhead : int;
  mutable stack_limit : int;
  mutable allocation_policy : int;
  window_size : int;
}

external stat : unit -> stat = "caml_gc_stat"
external quick_stat : unit -> stat = "caml_gc_quick_stat"
external counters : unit -> (float * float * float) = "caml_gc_counters"
external minor_words : unit -> (float [@unboxed])
  = "caml_gc_minor_words_unboxed"
external get : unit -> control = "caml_gc_get"
external set : control -> unit = "caml_gc_set"
external minor : unit -> unit = "caml_gc_minor"
external major_slice : int -> int = "caml_gc_major_slice"
external major : unit -> unit = "caml_gc_major"
external full_major : unit -> unit = "caml_gc_full_major"
external compact : unit -> unit = "caml_gc_compaction"
external get_minor_free : unit -> int = "caml_get_minor_free"
external get_bucket : int -> int = "caml_get_major_bucket" [@@noalloc]
external get_credit : unit -> int = "caml_get_major_credit" [@@noalloc]
external huge_fallback_count : unit -> int = "caml_gc_huge_fallback_count"

open Printf

let print_stat c =
  let st = stat () in
  fprintf c "minor_collections: %d\n" st.minor_collections;
  fprintf c "major_collections: %d\n" st.major_collections;
  fprintf c "compactions:       %d\n" st.compactions;
  fprintf c "\n";
  let l1 = String.length (sprintf "%.0f" st.minor_words) in
  fprintf c "minor_words:    %*.0f\n" l1 st.minor_words;
  fprintf c "promoted_words: %*.0f\n" l1 st.promoted_words;
  fprintf c "major_words:    %*.0f\n" l1 st.major_words;
  fprintf c "\n";
  let l2 = String.length (sprintf "%d" st.top_heap_words) in
  fprintf c "top_heap_words: %*d\n" l2 st.top_heap_words;
  fprintf c "heap_words:     %*d\n" l2 st.heap_words;
  fprintf c "live_words:     %*d\n" l2 st.live_words;
  fprintf c "free_words:     %*d\n" l2 st.free_words;
  fprintf c "largest_free:   %*d\n" l2 st.largest_free;
  fprintf c "fragments:      %*d\n" l2 st.fragments;
  fprintf c "\n";
  fprintf c "live_blocks: %d\n" st.live_blocks;
  fprintf c "free_blocks: %d\n" st.free_blocks;
  fprintf c "heap_chunks: %d\n" st.heap_chunks


let allocated_bytes () =
  let (mi, pro, ma) = counters () in
  (mi +. ma -. pro) *. float_of_int (Sys.word_size / 8)


external finalise : ('a -> unit) -> 'a -> unit = "caml_final_register"
external finalise_last : (unit -> unit) -> 'a -> unit =
  "caml_final_register_called_without_value"
external finalise_release : unit -> unit = "caml_final_release"


type alarm = bool ref
type alarm_rec = {active : alarm; f : unit -> unit}

let rec call_alarm arec =
  if !(arec.active) then begin
    finalise call_alarm arec;
    arec.f ();
  end


let create_alarm f =
  let arec = { active = ref true; f = f } in
  finalise call_alarm arec;
  arec.active


let delete_alarm a = a := false
